airbyte-integrations/connectors/source-sendgrid/source_sendgrid/streams.py

```diff
@@ -0,0 +1,273 @@
+        if next_page_url:
+            return {"next_page_url": next_page_url.replace(self.url_base, "")}
+
+    @staticmethod
+    @abstractmethod
+    def initial_path() -> str:
+        """
+        :return: initial path for the API endpoint if no next metadata url found
+        """
+
+    def path(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> str:
+        if next_page_token:
+            return next_page_token["next_page_url"]
+        return self.initial_path()
+
+
+class Scopes(SendgridStream):
+    def path(self, **kwargs) -> str:
+        return "scopes"
+
+
+class Lists(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/lists"
+
+
+class Campaigns(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/campaigns"
+
+
+class Contacts(SendgridStream):
+    data_field = "result"
+
+    def path(self, **kwargs) -> str:
+        return "marketing/contacts"
+
+
+class StatsAutomations(SendgridStreamMetadataPagination):
+    data_field = "results"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/stats/automations"
+
+
+class Segments(SendgridStream):
+    data_field = "results"
+
+    def path(self, **kwargs) -> str:
+        return "marketing/segments"
+
+
+class SingleSends(SendgridStreamMetadataPagination):
+    """
+    https://docs.sendgrid.com/api-reference/marketing-campaign-stats/get-all-single-sends-stats
+    """
+
+    data_field = "results"
+
+    @staticmethod
+    def initial_path() -> str:
+        return "marketing/stats/singlesends"
+
+
+class Templates(SendgridStreamMetadataPagination):
+    data_field = "result"
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        params["generations"] = "legacy,dynamic"
+        return params
+
+    @staticmethod
+    def initial_path() -> str:
+        return "templates"
+
+
+class Messages(SendgridStream, SendgridStreamIncrementalMixin):
+    """
+    https://docs.sendgrid.com/api-reference/e-mail-activity/filter-all-messages
+    """
+
+    data_field = "messages"
+    cursor_field = "last_event_time"
+    primary_key = "msg_id"
+    limit = 1000
+
+    def request_params(self, stream_state: Mapping[str, Any], **kwargs) -> MutableMapping[str, Any]:
+        time_filter_template = "%Y-%m-%dT%H:%M:%SZ"
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        if isinstance(params["start_time"], int):
+            date_start = datetime.datetime.utcfromtimestamp(params["start_time"]).strftime(time_filter_template)
```

Comment:
I hate adding a change in behavior into what is supposed to be a revert, but I can't get the the tests run from the GH action to pass w/o this. I'm not really sure how this passed prior to this change, but this will at least guarantee a consistent experience for everyone